### PR TITLE
lib/types: add `record` (simplified submodule)

### LINKED ISSUE #158598: `config`-driven submodules?
This assumes #158594 is implemented as an incremental change to the module system. I'm writing this as a separate issue to manage the scope; feel free to ignore this one.

For modules with large amounts of unset options, it is inefficient for module system evaluation to be guided by the opts, as the defs are significantly smaller value. In its early history, module system evaluation primarily traversed `config`, but presumably, this was abandoned with the introduction of defaults in options.

Making this mode of module evaluation available on an opt-in basis would allow large, frequently used submodules such as the ones in `systemd` to be modeled more efficiently, providing documentation, type checking and merging for this important part of NixOS.

### Notify maintainers

@Infinisil

### PR DESCRIPTION
**Pull request by MattSturgeon**
## Description of changes

Work in progress draft, based on #257511 by @roberth, but without `types.fix` and with additional support for optional fields.

Pushing an early draft, I'll polish up the changes and this description shortly.

The main TODO for me is to add additional test-cases to cover more exotic useage. I've added a few inline TODO comments regarding things I'm unsure of or plan to investigate.

More TODO from https://github.com/NixOS/nixpkgs/pull/257511/files#r1621120811:
- required fields can be returned without looking at their declarations, making it a bit more lazy, which is good for performance and robustness
- if we don't have optional or wildcard fields, we can return all the fields without even looking at the definitions
- it avoids having to partition the required and optional fields at runtime, which we'd then have to do; it's more efficient this way

---

- Fixes #158598
- Closes #333799
- Based on #257511
- See also #158594

cc @roberth @infinisil @GaetanLepage @traxys

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by aanderse** [comment]
can we nest records? something like this?

```
settings = record {
  fields {
    bar = int
  }
  optionalFields {
    foo = record {
      fields {
        a = int
        b = ...
      }
    }
  }
}
```

**Comment by MattSturgeon** [comment]
> can we nest records?

Yes. Currently nesting records (or submodules) within a record is the only way to achieve nested options.

So you can't do (for example):

```nix
settings = record {
  fields {
    foo.bar = mkOption {};
  };
}
```

But you can do:

```nix
settings = record {
  fields {
    foo = mkOption {
      type = record {
        fields = {
          bar = mkOption {};
        };
      };
    };
  };
}
```

This limitation was a design decision by @roberth in the original PR #257511, for performance reasons. I suppose it could be re-evaluated in the future though.

> something like this?

Almost. All fields are options, not types.

**Comment by aanderse** [comment]
thanks for the answer

i think something like this is an important but missing feature currently
i will stay subscribed to this PR and look forward to any updates if you get around to it

> All fields are options, not types.

of course, i was just simplifying

**Comment by roberth** [comment]
I'm in favor. Besides my earlier reasons, we've seen that [taggedSubmodule](https://github.com/NixOS/nixpkgs/pull/254790) is rather hard to implement in a complete way, whereas a solution based on `record` would be simpler, and sufficient for its purposes.

Probably it can even reuse `record` in whole - not duplicating its implementation. Its methods can easily retrieve the `type` value because definitions for `record` have no module arguments, fix-point, or unnecessary syntax, and when it knows the `type` value, it can get the right `record` type, merge a (`raw`) `type` field into that, and defer merging to this combined type. (That was probably hard to follow because everything is called "type" - I swear I'm not talking gibberish)

Anyway, I'd like to hear @infinisil's thoughts too.

**Review by roberth — COMMENTED** [review]
Thoughts and possible suggestions (if they're good; this is all very new)

File: lib/types/record.nix [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+          throw "Record field `${lib.escapeNixIdentifier name}` must be declared with `mkOption`."
+        else if value ? apply then
+          throw "In field ${lib.escapeNixIdentifier name} records do not support options with `apply`"
+        else if value ? readOnly then
+          throw "In field ${lib.escapeNixIdentifier name} records do not support options with `readOnly`"
+        else
+          value
+      );
+      fields' = checkFields fields;
+      # TODO: should we also check optionalFields have no default?
```
[the remark this answers is not in the document]
Comment (MattSturgeon) [reply lib/types/record.nix:36]: I agree a `mkField` would make sense to further de-couple records from modules. This would also allow fields to declare whether they are optional, rather than having a separate `optionalFileds` argument.

File: lib/types/record.nix [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+          optionalFieldValues = lib.concatMapAttrs (
+            fieldName: fieldOption:
+            builtins.addErrorContext
+              "while evaluating the optional field `${fieldName}' of option `${showOption loc}'"
+              (
+                let
+                  opt = mergeDefinitions (loc ++ [ fieldName ]) fieldOption.type (
+                    data.${fieldName} or [ ]
+                    # TODO: surely defaults make no sense for optional fields?
+                    # And doesn't mergeDefinitions do this internally, anyway?
```
[the remark this answers is not in the document]
Comment (MattSturgeon) [reply lib/types/record.nix:87]: It sounds like we need a lazy and non-lazy mode for record types, since its users may wish for either behaviour.

In nixvim, for instance, we probably want the non-lazy behaviour where optional fields are _not_ present in the resulting attrs.

This is because we usually want to be able to recursively map the attrs into a lua table using our version of `lib.generators.toLua`.

I guess we could have our `toLua` generator wrap recursion in `tryEval` and not render anything that throws, but that seems like an overly large hammer for this issue.

File: lib/types/record.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (MattSturgeon) [reply lib/types/record.nix]: I added `lib.fields` (`lib/fields.nix`) to host a `mkField` function, although maybe it'd be better to just have `mkField` in `lib/options.nix`?

**Comment by MattSturgeon** [comment]
This is still far from ready and I don't have _much_ time to dedicate to it.
However, I've made a few refactors and I would appreciate some feedback on the current design & direction.

(ping @roberth @aanderse)

**Comment by aanderse** [comment]
i keep seeing (and writing) more nix code that would benefit from this so much - these `settings` type options we use in nixos (and beyond) really need something like this... and the documentation for this looks pretty good to me

cc @artemist because of how much they [could benefit from this](https://github.com/nixos-bsd/nixbsd/blob/8acfedff2318f4abaa99bdd2bd8cb3505ce3d854/modules/config/user-class.nix#L99)

**Comment by hsjobeki** [comment]
I guess then i'm also interested in this PR.

Came over from: https://github.com/NixOS/nixpkgs/pull/377718

I'll check if i find some time helping out.

File: lib/fields.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      # Example value used in the manual.
+      example ? null,
+      # String describing the field.
+      description ? null,
+      # Related packages used in the manual (see `genRelatedPackages` in ../nixos/lib/make-fields-doc/default.nix).
+      relatedPackages ? null,
+      # Option type, providing type-checking and value merging.
+      type ? null,
+      # Whether the field is for NixOS developers only.
+      internal ? null,
```
Comment (roberth) [inline lib/fields.nix:26]:
```suggestion
```

I don't think we've ever made us of the difference between `internal` and `visible`, and `visible` is the more capable one, supporting `"shallow"`.

File: lib/tests/modules.sh [hunk 4]
```diff
@@ -107,6 +107,43 @@ checkConfigError() {
...
+# record optional field without definition
+checkConfigError "attribute 'age' in selection path 'config.people.alice.age' not found" config.people.alice.age ./declare-record-optional-field.nix ./define-record-alice.nix
+# record optional field with definition
+checkConfigOutput '^27$' config.people.mike.age ./declare-record-optional-field.nix ./define-record-mike.nix
+
+#TODO:
+# - test empty definitions
+# - test neseted records
+# - test nested optional records
+# - etc?
```
Comment (roberth) [inline lib/tests/modules.sh:142]: Docs generation (in `tests/misc.nix`?)

File: lib/tests/modules.sh [hunk 5]
```diff
@@ -107,6 +107,43 @@ checkConfigError() {
...
+
+
+if false; then
```
Comment (roberth) [inline lib/tests/modules.sh:145]:
```suggestion
```

File: lib/types/record.nix [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    ;
+
+  inherit (lib.strings)
+    escapeNixIdentifier
+    ;
+
+  record =
+    {
+      fields ? { },
+      freeformType ? null,
```
Comment (roberth) [inline lib/types/record.nix:29]: The freeform type applies at the root of a submodule / configuration, whereas this one applies to individual field values; not the attribute set containing them. I think that's a significant difference that should be reflected in the name. I think it was `wildcardType`.

File: nixos/doc/manual/development/freeform-modules.section.md [hunk 7]
```diff
@@ -12,6 +14,43 @@ into the resulting `config` set. Since this feature nullifies name
 checking for entire option trees, it is only recommended for use in
 submodules.
 
+<!-- TODO: Link to more details on the record and submodule types -->
+
+Freeform records can also be used to achieve the same thing, and may be
+preferred in many scenarios due to their improved performance when
+compared to submodules and also the ability to declare "optional" fields.
```
Comment (roberth) [inline nixos/doc/manual/development/freeform-modules.section.md:21]:
```suggestion
A similar result can be achieved with `types.record` with a `wildcardType`. Whereas `submodule` is designed to host logic inside of it, `record` is just a data container, and that lets it implement optional fields.
```

File: nixos/doc/manual/development/option-types.section.md [hunk 8]
```diff
@@ -211,9 +211,24 @@ merging is handled.
     definitions cannot be merged. The regular expression is processed
     using `builtins.match`.
 
-## Submodule types {#sec-option-types-submodule}
+## Record and module types {#sec-option-types-submodule}
```
Comment (roberth) [inline nixos/doc/manual/development/option-types.section.md:214]: I feel like `record` is closer to `attrTag`, `listOf` and `attrsOf`, but that might be overly theoretical of me.

File: lib/fields.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+      # Default value used when no definition is given in the configuration.
+      default ? null,
+      # Textual representation of the default, for the manual.
+      defaultText ? null,
+      # Example value used in the manual.
+      example ? null,
+      # String describing the field.
+      description ? null,
+      # Related packages used in the manual (see `genRelatedPackages` in ../nixos/lib/make-fields-doc/default.nix).
+      relatedPackages ? null,
```
Comment (roberth) [inline lib/fields.nix:22]: thought:
These could be bundled into a single `meta` attribute, and that might further improve performance by virtue of not being turned into values during non-docs evaluation.
I don't know how significant that is, and it may make `mkField` annoying to use, typing more and being dissimilar from `mkOption`.
Also the performance improvement could be inverted when any of the attrs inside of `meta` were to somehow become part of the main evaluation later (unlikely?).

File: nixos/doc/manual/development/option-types.section.md (hunk 8 above)
Comment (MattSturgeon) [reply nixos/doc/manual/development/option-types.section.md:214]: I agree records are technically dissimilar to submodules, but I think the usage of modules and records _is_ similar.

I interpret this section of the docs as laying out the available types for scenarios involving sub-options, and scoped/nested config.

**Comment by aanderse** [comment]
this PR remains one of the only viable proposed solution to the problems presented by RFC42

> I'll check if i find some time helping out.

as one of the more qualified people to work on the module system are you able to help out @hsjobeki? how about you @MattSturgeon, is this something that can come back on your radar?

any and all effort to push this through would be so greatly appreciated by myself, and many others i am sure :bowing_man:

**Comment by hsjobeki** [comment]
@MattSturgeon i hope its okay that i rebased your commits. I'll look into the reasons, why this is still not merged yet.

- merge should implement v2 ? does it even need a v1 ?
- freeformType, should be named `wildcard` or similar, to prevent confusion. Maybe we could omit this feature in the first iteration.
- Some more test seem to be missing (see modules.sh)

Otherwise, i think this looks pretty good already.

I would love to help out where i can :+1:

**Comment by MattSturgeon** [comment]
If course! Any help is appreciated. Are you at NixCon?

I've let this sit long enough that I'll need to sit down and re-read it before continuing. IIRC the biggest blocker was finding inspiration to write new test cases.

If you're motivated to work on this, you're welcome to take over. Or with clear/frequent feedback I'll be more motivated to work on this.

**Comment by hsjobeki** [comment]
> If course! Any help is appreciated. Are you at NixCon?
 I'm around, mostly in building 4

**Comment by hsjobeki** [comment]
After looking into this for a while the biggest foreseeable todos are

- Documentation generation. We need to turn a `mkField` into an artifical option for `optionAttrSetToDocList'` or look for another docs solution. Otherwise records will not have documentation fields. Native support for fields would be nice but requires changing a lot of tooling, such as search.nixos.org and probably the database thats running behind it.

- A bit more Testing

**Comment by roberth** [comment]
> * Documentation generation. We need to turn a `mkField` into an artifical option for `optionAttrSetToDocList'` or look for another docs solution

We chatted a bit about this at NixCon. We already have multiple representations for things that "are" options in one way or another. Allowing fields to be represented in the documentation-flavored option representations sounds more than ok to me. Ultimately the UX is pretty much the same, whether it was originally a field or `mkOption` option, so it makes sense to union them into a single documentation-oriented representation. In practice I think this means mostly transforming field documentation into option documentation, and that's ok.

**Comment by hsjobeki** [comment]
Fiy i also tried adding `optional` to mkOption. Just out of curiosity to explore the solution space. Which kind of works, but i couldn't find a way to not break lazyness of options.
I need to compare how records handle that in comparison. For example lazyAttrs

**Comment by hsjobeki** [comment]
@roberth since we need to move `optional` itself out of `mkField` 
Should we use `mkOption` instead ? (Would reduce the amount of things added, and avoid adding a new term)

But this also requires handling of  `readOnly` and `apply`

Those are handled by `evalOptionValue` not sure if we want that. We could probably also use `mergeModules` (or `evalOptionValue` or handle everything ourselves in the `merge` function of the record type?

Or we could explicitly exclude the apply and readonly features? I would expect at least readOnly to remain supported?

**Comment by Mic92** [comment]
I measured the performance impact on NixOS if we had such a type: https://github.com/NixOS/nixpkgs/pull/509197
